File: Weekly Contest 223 (0110)/4/5639_1723_H_FindMinimumTimeFinishAllJobs.js
// https://leetcode.com/contest/weekly-contest-223/problems/find-minimum-time-to-finish-all-jobs/
const getMax = (arr, n) => {
    let result = arr[0];
    for (let i = 1; i < n; i++)
        if (arr[i] > result)
            result = arr[i];
    return result;
}

const isPossible = (time, K, job, n) => {
    let cnt = 1;
    let curr_time = 0;
    for (let i = 0; i < n;) {
        if (curr_time + job[i] > time) {
            curr_time = 0;
            cnt++;
        }
        else {
            curr_time += job[i];
            i++;
        }
    }
    return (cnt <= K);
};

const minimumTimeRequired = (job, K) => {
    let n = job.length;
    let end = start = 0;
    for (let i = 0; i < n; ++i) {
        end += job[i];
    }
    let ans = end;
    let job_max = getMax(job, n);
    while (start <= end) {
        let mid = (start + end) / 2;
        if (mid >= job_max && isPossible(mid, K, job, n)) {
            ans = Math.min(ans, mid);
            end = mid - 1;
        } else {
            start = mid + 1;
        }
    }
    return ans;
};

const main = () => {
    let jobs = [3, 2, 3], k = 3;
    let jobs2 = [1, 2, 4, 7, 8], k2 = 2;
    let job3 = [10, 7, 8, 12, 6, 8], k3 = 3;
    console.log(minimumTimeRequired(jobs, k));
    console.log(minimumTimeRequired(jobs2, k2));
    console.log(minimumTimeRequired(job3, k3));
};

main()

File: 2021/Weekly Contest 223 (0110)/4/5639_1723_H_FindMinimumTimeFinishAllJobs.py
"""
https://www.tutorialspoint.com/find-minimum-time-to-finish-all-jobs-with-given-constraints-in-python
"""

# WA
class Solution(object):
    def is_valid(self, time, K, job):
        n = len(job)
        count = 1
        curr_time = 0
        i = 0
        while i < n:
            if curr_time + job[i] > time:
                curr_time = 0
                count += 1
            else:
                curr_time += job[i]
                i += 1
        return count <= K

    def minimumTimeRequired(self, job, K):
        n = len(job)
        end = 0
        begin = 0
        for i in range(n):
            end += job[i]
        res = end
        job_max = max(job)
        while begin <= end:
            print(res)
            mid = int((begin + end) >> 1)
            if mid >= job_max and self.is_valid(mid, K, job):
                res = min(res, mid)
                end = mid - 1
            else:
                begin = mid + 1
        return res


if __name__ == "__main__":
    s = Solution()
    jobs = [3, 2, 3]
    k = 3
    jobs2 = [1, 2, 4, 7, 8]
    k2 = 2
    print(s.minimumTimeRequired(jobs, k))
    print(s.minimumTimeRequired(jobs2, k2))
